### PR TITLE
Split broccoli-middleware into two different middlewares

### PR DESCRIPTION
**Pull request by kratiahuja**
Implementation of [ember-cli RFC # 80](https://github.com/ember-cli/rfcs/pull/80). This PR is only the `broccoli-middleware` side of changes.

### CODE REVIEW

**Comment by jfelchner** [comment]
I think this is going to help with some problems I've been having with the errors page. ☺️

**Comment by kratiahuja** [comment]
@rwjblue should we update appveyor to not run Node 0.10 version since we dropped it in travis? I have a PR for this [here](https://github.com/ember-cli/broccoli-middleware/pull/22).

Tests pass otherwise.

File: lib/index.js
Comment (hjdivad) [inline lib/index.js:9]: What's the motivation for retaining a combined middleware?

File: lib/watcher-middleware.js
Comment (hjdivad) [inline lib/watcher-middleware.js]: A couple of points here:

1. I'm surprised to see so much of the response header set here.  eg `setResponseHeaders` sets things like `content-length` which seems to be a concern of the `serveAssetMiddleware` not the `watcherMiddleware`.  In [this comment](https://github.com/ember-cli/rfcs/pull/80#pullrequestreview-9709454) the only headers set up in the watcher middleware were the `x-broccoli` request header for tracking the output path and the `Cache-Control` response header as that's a pretty reasonable header to set as a default irrespective of how the request is actually fulfilled.

Why are file-specific headers set in this middleware rather than the asset serving middleware?

2. It seems weird for `setResponseHeaders` to not be a middleware but to still accept `next`.  I'm not suggesting that it should be a middleware: I'm worried that it's currently hiding some of the control flow.  Is there some way we can split up this function?

File: lib/watcher-middleware.js
Comment (hjdivad) [inline lib/watcher-middleware.js]: In what ways can the file name be updated?  This code looks general and the comment makes it sound general, but from what I could tell of `setResponseHeaders` the only change is that `index.html` may be appended.  Did I miss something?

File: lib/serve-asset-middleware.js
Comment (hjdivad) [inline lib/serve-asset-middleware.js:24]: What's the use case for `noop`ing here?

Comment (kratiahuja) [reply lib/serve-asset-middleware.js:24]: In `ember-cli` the last middleware in the chain is the [proxy-server-middleware](https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/server/middleware/proxy-server/index.js) which is invoked when a file is not found in disk or if index.html and autoIndex is not true. Basically the case when someone uses `ember serve --proxy http://localhost:3000` for example. In this case, when we know the file is not found we want to bypass this middleware (ie not even serve it) and let the proxy server handle the request.

File: lib/index.js
Comment (kratiahuja) [reply lib/index.js:9]: It was more for maintaining backward compatibility. If we need to do a point release of ember-cli that is using the combined middleware and requires a bug fix in broccoli-middleware.

File: lib/serve-asset-middleware.js
Comment (hjdivad) [reply lib/serve-asset-middleware.js:24]: 👍

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: 1. I am not adding to set any extra headers. These headers were also set previously. I think the comment was more of an example to explain the workflow. The reason file specific headers are set in the middleware is because we want to provide these as the default implementation and let anyone override it downstream if needed rather than anyone who wants to insert itself before the asset serving middleware to almost copy over those header settings. 

2. I thought about splitting it but it wasn't very clean since it has dependencies on `stat` object and other things. I could do a small refactor such that it doesn't take `next`.

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: It can be updated when `index.html` is appended. Nope you didn't miss anything.

File: lib/index.js
Comment (hjdivad) [reply lib/index.js:9]: Seems reasonable.  Personally I would not include it in `master` but instead create a minor version branch if a patch version became necessary.  Not sure what the maintainer preference here is though.  Seems like a decision for @joliss @rwjblue @stefanpenner

File: lib/watcher-middleware.js
Comment (hjdivad) [reply lib/watcher-middleware.js]: > I am not adding to set any extra headers. These headers were also set previously. I think the comment was more of an example to explain the workflow. The reason file specific headers are set in the middleware is because we want to provide these as the default implementation and let anyone override it downstream if needed rather than anyone who wants to insert itself before the asset serving middleware to almost copy over those header settings.

Ah sorry I did not make my point very clearly.  I totally agree with all the headers: what I'm surprised about is that they're set in the watcher middleware and not the serve asset middleware.

> I thought about splitting it but it wasn't very clean since it has dependencies on stat object and other things. I could do a small refactor such that it doesn't take next.

Seems reasonable although this is relatively minor to me in comparison to figuring out why we're doing this in this middleware rather than serve assets.

File: lib/watcher-middleware.js
Comment (hjdivad) [reply lib/watcher-middleware.js]: Okay cool, do we need this in the watcher middleware? This is related to my point above, where it feels to me like this stuff is the concern of the asset serving middleware rather than the watcher middleware.

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: > Ah sorry I did not make my point very clearly. I totally agree with all the headers: what I'm surprised about is that they're set in the watcher middleware and not the serve asset middleware.

I would disagree here since there is a usecase where this will not work as expected (if we have serve asset middleware handler those headers). The purpose of the watcher middleware was to set all the generic headers and the serve file asset middleware to actually just serve from disk. The reason being: Imagine usecase where someone wants to insert a middleware between the watcher and serve asset middleware in order to serve assets themselves. If the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers. IMO that is undesirable. The custom middleware or any downstream middleware should only overwrite the headers.

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: Please see my response above.

File: lib/index.js
Comment (kratiahuja) [reply lib/index.js:9]: Yeah I am fine any path. I don't have much experience on how it works in ember-cli.

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: > Seems reasonable although this is relatively minor to me in comparison to figuring out why we're doing this in this middleware rather than serve assets.

I went ahead and refactored the code a bit so that `next` is not leaked into the utility function.

Comment (hjdivad) [reply lib/watcher-middleware.js]: > The purpose of the watcher middleware was to set all the generic headers and the serve file asset middleware to actually just serve from disk

I would have imagined the purpose of the watcher middleware to just be the error handling and to block the request until the watcher has finished.

> The reason being: Imagine usecase where someone wants to insert a middleware between the watcher and serve asset middleware in order to serve assets themselves. If the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers.

So a couple of things:

1. If someone else is serving assets presumably there's something different about them, which is why they can't just fall back to the asset serving middleware.  This means that at a minimum headers like [Content-Length](https://github.com/ember-cli/broccoli-middleware/blob/aac8215f5184b226c51e964313c5c49460cf0da1/lib/utils/response-header.js#L116) and [Last-Modified](https://github.com/ember-cli/broccoli-middleware/blob/aac8215f5184b226c51e964313c5c49460cf0da1/lib/utils/response-header.js#L101) are likely to be wrong.

2. It seems weird that they can only sometimes rely on the default headers.  The headers set in `setResponseHeaders` depend on being able to actually find the file on disk.  Someone who wants to write a middleware between the watcher and asset serving may or may not actually be serving assets from disk or not.  The reason this isn't a problem with the asset serving middleware is that serving files found on disk takes precedence over the proxy, but for someone who wanted to insert their middleware *before* asset serving, they would want their middleware to take precedence.

> if the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers.

Maybe we could move that logic into `lib/utils/serve-asset.js` and make it public?  This way it would be still be pretty easy to write a middleware that wanted to serve a static asset just like serve assets, but whose url -> file on disk algorithm was different (eg as with fastboot which wants `/` -> `/index.fastboot.html` instead of `/` -> `/index.html` or something similar)

Comment (kratiahuja) [reply lib/watcher-middleware.js]: Sure we can do that. I'll update this PR.

**Comment by kratiahuja** [comment]
@hjdivad Updated per your suggestion. Please let me know if I misunderstood anything.

File: lib/serve-asset-middleware.js
Comment (hjdivad) [inline lib/serve-asset-middleware.js]: I don't think `serveAssetsMiddleware` needs to be a function that returns a middleware.  We can just export the middleware directly.

The reason the previous implementation had a function that returned a middleware was to close over the `watcher` argument, but that's not necessary here.

ie we can just `module.exports = function(request, response, next) { // ...`

File: lib/utils/response-header.js
Comment (hjdivad) [inline lib/utils/response-header.js]: This line seems to be a no-op; if the above conditional matches then `req.headers['x-broccoli'] == broccoliInfo` already

File: lib/watcher-middleware.js
Comment (hjdivad) [inline lib/watcher-middleware.js]: dead code

File: test/middleware-test.js
Comment (hjdivad) [inline test/middleware-test.js:113]: 👍 

Always good to make tests tell the truth 😄

File: test/serve-asset-middleware-test.js
Comment (hjdivad) [inline test/serve-asset-middleware-test.js:62]: What does the wrapper achieve?  Can't we just `server.addMiddleware(serveAssetMiddleFn)`

File: test/serve-asset-middleware-test.js
Comment (hjdivad) [inline test/serve-asset-middleware-test.js:91]: as above, unclear on what we get out of the wrapper here

File: test/serve-asset-middleware-test.js
Comment (hjdivad) [inline test/serve-asset-middleware-test.js]: as above, unsure about the wrapper

File: test/watcher-middleware-test.js
Comment (hjdivad) [inline test/watcher-middleware-test.js:41]: what's the point of the wrapper here?  can't we use `middleware` directly?

**Review by hjdivad — APPROVED** [review]
I have a number of very minor nits but nothing substantial.  I think it would be fine to fix them before or after merge.  High level everything looks great.

Thanks @kratiahuja  👍

File: lib/serve-asset-middleware.js
Comment (kratiahuja) [reply lib/serve-asset-middleware.js]: I was just following the format of other middlewares for consistency. Fixed to just not have the closure 😄

File: lib/utils/response-header.js
Comment (kratiahuja) [reply lib/utils/response-header.js]: Sorry about that. Fixed.

File: lib/watcher-middleware.js
Comment (kratiahuja) [reply lib/watcher-middleware.js]: Fixed.

File: test/serve-asset-middleware-test.js
Comment (kratiahuja) [reply test/serve-asset-middleware-test.js:62]: The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after `serveAssetMiddleware` is finished executing.

File: test/serve-asset-middleware-test.js
Comment (kratiahuja) [reply test/serve-asset-middleware-test.js:91]: The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after serveAssetMiddleware is finished executing.

File: test/serve-asset-middleware-test.js
Comment (kratiahuja) [reply test/serve-asset-middleware-test.js]: The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after serveAssetMiddleware is finished executing.

File: test/watcher-middleware-test.js
Comment (kratiahuja) [reply test/watcher-middleware-test.js:41]: The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after serveAssetMiddleware is finished executing.

**Comment by kratiahuja** [comment]
@hjdivad Thanks for your feedback. I apologize for the delay in updating the PR. I believe I have addressed the nitpicks too with the latest push. Could you please take another quick look? 😄 

If this is good, could we please merge it in (only `master`)?

**Comment by hjdivad** [comment]
Looks great, thanks @kratiahuja!

**Comment by kratiahuja** [comment]
@hjdivad @stefanpenner @rwjblue  need another favor. Can we do a release of `broccoli-middleware` from `master` as `1.0.0-beta8`? I need this to get tests passing in `ember-cli`.

**Comment by stefanpenner** [comment]
releasing now

**Comment by stefanpenner** [comment]
should be released, let me know if there is an issue.

**Comment by kratiahuja** [comment]
Thank you @stefanpenner ! I'll let you know if I hit any issues.

**Pull request merged**
